contracts/epochs/FlowEpoch.cdc

```diff
@@ -119,36 +131,458 @@ pub contract FlowEpochs {
+        let nodeInfo = FlowIDTableStaking.NodeInfo(nodeID: nodeStaker.id)
+
+        assert (
+            nodeInfo.role == 1 as UInt8,
+            message: "Node operator must be a collector node to get a QC Voter object"
+        )
+
+        let clusterQCAdmin = self.borrowClusterQCAdmin()
+
+        return <-clusterQCAdmin.createVoter(nodeID: nodeStaker.id)
+    }
+
+    pub fun getDKGParticipant(nodeStaker: &FlowIDTableStaking.NodeStaker): @FlowDKG.Participant {
+        let nodeInfo = FlowIDTableStaking.NodeInfo(nodeID: nodeStaker.id)
+
+        assert (
+            nodeInfo.role == 2 as UInt8,
+            message: "Node operator must be a consensus node to get a DKG Participant object"
+        )
+
+        let dkgAdmin = self.borrowDKGAdmin()
+
+        return <-dkgAdmin.createParticipant(nodeID: nodeStaker.id)
+    }
+
+    pub fun createCollectorClusters(nodeIDs: [String]): [FlowEpochClusterQC.Cluster] {
+        var shuffledIDs = self.randomize(nodeIDs)
+
+        // Holds cluster assignments for collector nodes
+        let clusters: [FlowEpochClusterQC.Cluster] = []
+        var clusterIndex: UInt16 = 0
+        let nodeWeightsDictionary: [{String: UInt64}] = []
+
+        for id in shuffledIDs {
+
+            let nodeInfo = FlowIDTableStaking.NodeInfo(nodeID: id)
+
+            if nodeWeightsDictionary[clusterIndex] == nil {
+                nodeWeightsDictionary[clusterIndex] = {}
+            }
+
+            nodeWeightsDictionary[clusterIndex][id] = nodeInfo.initialWeight
+            
+            // Advance to the next cluster, or back to the first if we have gotten to the last one
+            clusterIndex = clusterIndex + 1 as UInt16
+            if clusterIndex == self.numCollectorClusters {
+                clusterIndex = 0
+            }
+        }
+
+        // Create the clusters Array that is sent to the QC contract
+        // and emitted in the EpochSetup event
+        clusterIndex = 0
+        while clusterIndex < self.numCollectorClusters {
+            clusters.append(FlowEpochClusterQC.Cluster(index: clusterIndex, nodeWeights: nodeWeightsDictionary[clusterIndex]!))
         }
+
+        return clusters
+
     }
+  
+    /// A function to generate a random permutation of arr[] 
+    /// using the fisher yates shuffling algorithm
+    pub fun randomize(_ array: [String]): [String] {  
+
+        var i = array.length - 1
+
+        // Start from the last element and swap one by one. We don't 
+        // need to run for the first element that's why i > 0 
+        while i > 0
+        { 
+            // Pick a random index from 0 to i 
+            var randomIndex = Int(unsafeRandom()) % (i + 1)
+    
+            // Swap arr[i] with the element at random index 
+            var temp = array[i]
+            array[i] = array[randomIndex]
+            array[randomIndex] = temp
+
+            i = i - 1
+        }
+
+        return array
+    } 
+
+
+    init (numViewsInEpoch: UInt64, 
```

Comment:
The counter for the first epoch should be an argument here -- we need to be able to synchronize that between the protocol state and here when we're initializing.

We should also accept the `NodeInfo` list that was generated in the protocol bootstrapping, then we can fail if it doesn't match what we see in the staking contract.

More generally, we should accept each field of the `Setup` and `Commit` events here (even the ones that are usually computed like the DKG phase deadlines) and use those inputs as the source of truth for the first epoch. If the inputs are inconsistent (I think the only time this could happen is inputs related to the identity table which we need to check against the staking contract) then we should abort.

This is what I think the bootstrapping process should look like:
1. `Protocol Bootstrap Script` retrieves list of nodes for next epoch from staking contract
2. `Protocol Bootstrap Script` generates the root state for the spork using the node list from step 1 (this root state includes the `EpochSetup` and `EpochCommit` service events for the first post-spork epoch
3. `Protocol Bootstrap Script` initializes `FlowEpoch` contract, passing in the service events generated in step 2
4. `FlowEpoch` contracts initializes its internal state such that:
  * `currentEpochCounter` is the counter for the input service events
  * `currentEpochPhase` is `STAKING`
  * Epoch metadata for the current epoch is consistent with the input service events

Reply:
Do we actually need all the inputs for the service events in `init`? We decided that we aren't emitting the events in `init`, correct? So all we need is what is stored in `EpochMetadata`.

Reply:
Yes good point. We only need what is stored in `EpochMetadata`